EIPS/eip-5593.md

```diff
@@ -0,0 +1,75 @@
+description: Wallet guidance for restricting web3 provider object API access to secure contexts for improved privacy and security for wallet users.
+author: Yan Zhu (@diracdeltas), Brian R. Bondy (@bbondy), Kyle Den Hartog (@kdenhartog)
+discussions-to: https://ethereum-magicians.org/t/rfc-limiting-provider-object-injection-to-secure-contexts/10670
+status: Draft
+type: Standards Track
+category: Interface
+created: 2022-09-05
+---
+
+## Abstract
+
+Historically the web platform has had a notion of “powerful” APIs like [geolocation](https://www.w3.org/TR/geolocation/) and [camera/microphone](https://www.w3.org/TR/mediacapture-streams/), which are subject to additional security restrictions such as those defined by [secure contexts](https://www.w3.org/TR/secure-contexts/). Since the web3 provider object APIs allow dApp websites to request access to sensitive user data and to request use of user funds, new web3 provider object APIs generally should align to the security considerations of other powerful APIs like geolocation.
+
+## Motivation
+
+Wallets are oftentimes maintaining security and safety of users' funds that can be equivalent to large portions of money. For this reason, it's a good idea to restrict access to the APIs that grant this access in certain contexts to align it more closely with the principles followed for web platform APIs given web3 provider object APIs are often extending this platform via extensions. By adding in restrictions to where extensions inject their APIs, we're reducing the surface area that malicious web pages could fingerprint or falsely request access to the web3 provider object APIs to harm users by accessing their data or stealing their web3 assets.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Restrictions for providers
+
+The provider objects, e.g. `window.ethereum`, are not provided in all contexts. The following restrictions are REQUIRED for conformant wallets:
+
+- Provider objects SHOULD NOT be accessible in private (incognito) windows.
+- The origin MUST be a [secure origin](https://www.chromium.org/Home/chromium-security/prefer-secure-origins-for-powerful-new-features/#definitions) to have access to `window.ethereum`. This can be checked using `window.isSecureContext`, including inside iframes.
+    - Secure contexts include sites that are served from HTTPS but also HTTP `localhost`.
+- By default the web3 provider object APIs MUST NOT be exposed to 3p iframes.
+- `window.ethereum` MUST be `undefined` in an iframe where `window.isSecureContext` returns `false` in that iframe.
+- If the iframe is first-party to the top-level origin AND the `sandbox` attribute is set on the iframe, the provider object MUST be left `undefined`. If `sandbox="allow-same-origin"` attribute is set the provider object MAY be injected. 
```

Comment:
Good point, I think this should become a MUST. I'll double check later to make sure this isn't an issue with the secure contexts implementations that makes it a MAY or if it's just a mistake I made when writing it up.

Reply:
Figured out why I put this as a MAY and updated it since it wasn't well defined. The reasoning for the MAY was that in the third party iframe case it wouldn't make a difference, so rather than leave it as a MAY I updated it to a MUST and specified it to be around first party frames. Third party frames would still be updated based on the `allow` attribute and the permissions API.